sdks/java/core/src/main/java/org/apache/beam/sdk/schemas/transforms/SchemaTransformProvider.java

```diff
@@ -0,0 +1,70 @@
+/*
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package org.apache.beam.sdk.schemas.transforms;
+
+import java.util.List;
+import java.util.Optional;
+import org.apache.beam.sdk.annotations.Experimental;
+import org.apache.beam.sdk.annotations.Experimental.Kind;
+import org.apache.beam.sdk.annotations.Internal;
+import org.apache.beam.sdk.options.PipelineOptions;
+import org.apache.beam.sdk.schemas.Schema;
+import org.apache.beam.sdk.values.Row;
+
+/**
+ * Provider to create {@link SchemaTransform} instances.
+ *
+ * <p><b>Internal only:</b> This interface is actively being worked on and it will likely change as
+ * we provide implementations for more standard Beam transforms. We provide no backwards
+ * compatibility guarantees and it should not be implemented outside of the Beam repository.
+ */
+@Internal
+@Experimental(Kind.SCHEMAS)
+public interface SchemaTransformProvider {
+  /** Returns an id that uniquely represents this transform. */
+  String identifier();
+
+  /**
+   * Returns the expected schema of the configuration object. Note this is distinct from the schema
+   * of the transform itself.
+   *
+   * <p>Configurations should be forwards compatible. Fields added after this SchemaTransform is
+   * first released should always be nullable.
+   */
+  Schema configurationSchema();
+
+  /**
+   * Produce a SchemaTransform from transform-specific configuration object. Can throw a {@link
+   * InvalidConfigurationException} or a {@link InvalidSchemaException}.
+   */
+  SchemaTransform from(Row configuration);
```

Comment:
Do you think we could make this strongly typed, instead of using generic Rows? That was one shortcoming in the original SchemaIO interfaces imo. It would be nice to discourage the use of generic rows.

```suggestion
  SchemaTransform from(ConfigT configuration);
```

This could accept a generic class `ConfigT`, and we could require that it be a type with an inferred schema.

We'd have to re-think how `configurationSchema` works. Perhaps we'd instead require a method like `Class<ConfigT> getConfigClass()`, and we can look up that type in the Schema registry, when we're registering the SchemaTransformProvider implementations.

Reply:
@laraschmidt and I discussed this a bit offline. We were a little concerned that _only_ supporting inferred schemas could limit velocity (e.g. what if we want to use a logical type that can't currently be inferred, and it takes substantial design work to support inferring it). Instead it makes sense to go with the current approach, and add a `TypedSchemaTransform` on top of it (either in this PR, or in a follow-up).

Reply:
Thanks Brian. I just added TypsedSchemaTransformProvider + test into this change.